### PR TITLE
Change API from []byte to reflect.Value

### PR DESCRIPTION
**Pull request by GreenHorse**
Use reflect.Value high level for the package API, remove scanner package.
Add Addition and Set.
Update `README` to match changes.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -48,48 +55,65 @@ The initial goal is to support all the selectors the original jq command line su
...
+  },
+  Map:         map[string]interface{}{"f": []int{6, 7, 8}},
+  WithJSONTag: 9,
 }
 ```
 
 | syntax | value |
 | :--- | :--- |
 | .string | "a" |
-| .number| 1.23 |
```
Comment (minjatJ) [inline README.md:75]: Floats are not supported anymore ?

File: err.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
+package jq
+
+import "errors"
+
+var (
+	// ErrCannotSet indicates the value is not settable, it will happend
+	// when a value is passed by copy
+	ErrCannotSet = errors.New("Cannot set value")
+	// ErrUnknownField indicates that we trie to access a field not present in
```
Comment (minjatJ) [inline err.go]: `we try`

File: op.go [hunk 3]
```diff
@@ -15,42 +15,231 @@
...
 // Apply executes the transformation defined by OpFunc
-func (fn OpFunc) Apply(in []byte) ([]byte, error) {
+func (fn OpFunc) Apply(in reflect.Value) (reflect.Value, error) {
 	return fn(in)
 }
 
+func callChain(v reflect.Value, chainFun ...Op) (reflect.Value, error) {
+	for _, f := range chainFun {
+		var err error
+		v, err = f.Apply(v)
```
Comment (minjatJ) [inline op.go]: `err` should be declare at the beginning then `err != nil` condition can be merged with assignment.

File: op.go [hunk 4]
```diff
@@ -15,42 +15,231 @@
...
+
+	return func(in reflect.Value) (reflect.Value, error) {
+		for in.Kind() == reflect.Interface || in.Kind() == reflect.Ptr {
+			in = in.Elem()
+		}
+		switch in.Kind() {
+		case reflect.Map:
+			var err error
+			mapVal := in.MapIndex(reflect.ValueOf(key))
+			newMapVal := reflect.New(in.MapIndex(reflect.ValueOf(key)).Type())
```
Comment (minjatJ) [inline op.go]: Why not using `mapVal` here ? (`reflect.New(mapVal)...`)

File: struct_json.go [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+			f := v.Type().Field(i)
+			t := f.Tag.Get("json")
+			strs := strings.Split(t, ",")
+			if strs[0] != "" {
+				buf[strs[0]] = i
+			}
+		}
+		jsonNameToIdx.LoadOrStore(v.Type(), buf)
+	}
+	buf, _ := jsonNameToIdx.Load(v.Type())
```
Comment (minjatJ) [inline struct_json.go]: Please store `v.Type()` result

File: op.go [hunk 6]
```diff
@@ -15,42 +15,231 @@
...
+			if valRef.Kind() != reflect.Slice {
+				return reflect.Value{}, fmt.Errorf("Addition: In is slice but val is: %v", valRef.Kind())
+			}
+			if in.Type() != valRef.Type() {
+				return reflect.Value{}, fmt.Errorf("Addition: cannot add elem from slice of type %v to slice of type %v", valRef.Type(), in.Type())
+			}
+			in.Set(reflect.AppendSlice(in, valRef))
+			return callChain(in, chainFun...)
+		case reflect.Map:
+			if v, ok := val.(json.RawMessage); ok {
```
Comment (minjatJ) [inline op.go]: `case reflect.Slice` has similar code. We should implement a `valueFromJSON` method

File: op.go [hunk 7]
```diff
@@ -15,42 +15,231 @@
...
+			return reflect.Value{}, errors.New("cannot access name field on slice")
+		}
+		return reflect.Value{}, errors.New("key not found")
+	}
+}
+
+// Addition adds the val parameter to the provided interface{} (map/slice/struct)
+func Addition(val interface{}, chainFun ...Op) OpFunc {
+	valRef := reflect.ValueOf(val)
+
```
Comment (minjatJ) [inline op.go]: We should maybe precompute `valRef.Kind/Type`

File: op.go [hunk 8]
```diff
@@ -15,42 +15,231 @@
...
+			}
+			return callChain(in, chainFun...)
+		}
+		return reflect.Value{}, fmt.Errorf("Unsupported type (%v)", valRef.Type())
 	}
+}
+
+// Set change the val parameter to the provided interface{} (map/slice/struct)
+func Set(val interface{}, chainFun ...Op) OpFunc {
+	valRef := reflect.ValueOf(val)
```
Comment (minjatJ) [inline op.go:195]: ditto

File: parse.go [hunk 9]
```diff
@@ -34,29 +38,189 @@ func Must(op Op, err error) Op {
...
+		idx++
+	}
+	return strings.Join(selectors, "=") + selector, args, nil
+}
+
 // Parse takes a string representation of a selector and returns the corresponding Op definition
-func Parse(selector string) (Op, error) {
+// TODO: move the parsing logic to another package
+func Parse(selector string, args ...interface{}) (Op, error) {
+	log.SetFlags(log.Llongfile)
```
Comment (minjatJ) [inline parse.go]: Debug code ?

File: parse.go [hunk 10]
```diff
@@ -34,29 +38,189 @@ func Must(op Op, err error) Op {
...
+		return Set(arg, op...)
+	}, true
 }
 
-func parseArray(key string) (Op, bool) {
+func parseAddition(key string, arg interface{}) (func(op ...Op) Op, bool) {
+	return func(op ...Op) Op {
+		return Addition(arg, op...)
+	}, true
+	// return nil, false
```
Comment (minjatJ) [inline parse.go]: dead code

**Review by minjatJ — CHANGES_REQUESTED** [review]
Nice work, only nit. Keep in mind that reflection is quite heavy so we should precompute as much as we can

File: op.go (hunk 4 above)
Comment (GreenHorse) [reply op.go]: No reason, it is even better for map of interfaces as we can obtain the concrete type for the key.

File: decode_go1.10.go [hunk 11]
```diff
@@ -0,0 +1,14 @@
+// +build go1.10
```
Comment (minjatJ) [inline decode_go1.10.go:1]: Good idea

**Review by minjatJ — APPROVED** [review]
Thanks for the changes. Nice one, LGTM

**Pull request merged**